### DISCUSSION TITLE
Questions about profiles/generations on MacOS

### DISCUSSION THREAD

**Discussion by ryanmsnyder** [dustinlyons/nixos-config#51] [Q&A] [answered]
A couple of questions relating to profiles/generations on MacOS:

1. How do I rollback to a previous generation? Do I need to use `./result/sw/bin/darwin-rebuild`?

2. This question is probably more geared toward `nix`/`nix-darwin` in general but why aren't `~/.nix-profile/Applications` and `/Applications/Nix Apps` always in step with each other? For example, when I:

- run `nix run .#build-switch`
- remove a package from the config
- run `nix run .#build-switch` again

that package remains in `~/.nix-profile/Applications` but is removed from `/Applications/Nix Apps`. I would expect it to be removed in `~/.nix-profile/Applications` as well because launchers such as Spotlight and Raycast still pick up the removed package/app since it's still in `~/.nix-profile/Applications`.

The second question may be completely unrelated to profiles/generations but I'm a nix noob and can't figure out why applications aren't being removed from `~/.nix-profile/Applications` when I remove them from my config.

Thank you for config. It's helping me learn a lot about nix.

**Comment by dustinlyons** [comment]
Hi @ryanmsnyder! Thanks for giving this project a try. Glad it's helping you learn Nix. I'm still always learning it ;)

In re: to your questions:

> How do I rollback to a previous generation? Do I need to use ./result/sw/bin/darwin-rebuild?

I added a new Nix app to the project for rollback. Check it out here:
https://github.com/dustinlyons/nixos-config/blob/main/templates/starter/apps/aarch64-darwin/rollback

If you just want to run the underlying commands and not add this into your configuration (assuming you already used the starter template), it's:

```
/run/current-system/sw/bin/darwin-rebuild --rollback
```

> This question is probably more geared toward nix/nix-darwin in general but why aren't ~/.nix-profile/Applications and /Applications/Nix Apps always in step with each other?

I don't have a quick answer for you. It depends on how you're installing the application. I know with the `homwbrew` module for `nix-darwin`, some apps aren't uninstalled automatically:

https://daiderd.com/nix-darwin/manual/index.html#opt-homebrew.masApps

I would expect the symlinks to be cleaned up, however. You should be able to manually clean it up yourself, however. That may be a workaround until you get a better answer. I would check with the `nix-darwin` folks.

**Reply by ryanmsnyder** [reply]
Thank you for your response @dustinlyons. I'm actually not using `homebrew` to install the applications. I'm just using `home-manager.users.${user}.home.packages`. 

One thing I just noticed in the config is that `modules/shared/packages.nix` is getting called in both:

- `modules/darwin/home-manager.nix` in `home-manager.users.${user}.home.packages` (user's profile)
- `hosts/darwin/default.nix` in `environment.systemPackages` (system-wide)

Is this intentional? That explains why the applications are getting symlinked in `~/.nix-profile/Applications` as well as `/Applications/Nix Apps`.

**Reply by ryanmsnyder** [reply, marked as answer]
Also, I've noticed the symlink _is_ being cleaned up from`~/.nix-profile/Applications` when I remove a package/app from being installed via `home-manager`. Likewise with executables in `~/.nix-profile/bin`. 

Finder/ForkLift was still showing the removed package in `~/.nix-profile/Applications` until I restarted them. I guess symlinks aren't automatically refreshed in Finder/ForkLift even if you navigate out of the directory containing the symlinks and then navigate back into it - they have to be relaunched/restarted.
